src/graph/serializers/encoder/with_server_events/v7/encode_graph_entities.c

```diff
@@ -0,0 +1,309 @@
+/*
+* Copyright 2018-2020 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#include <assert.h>
+#include "encode_v7.h"
+#include "../../../../../util/arr.h"
+#include "../../../../../datatypes/array.h"
+
+extern uint64_t entities_threshold;
+
+// Forward declerations.
+static void _RdbSaveSIArray(RedisModuleIO *rdb, const SIValue array);
+
+static void _RdbSaveSIValue(RedisModuleIO *rdb, const SIValue *v) {
+	/* Format:
+	 * SIType
+	 * Value */
+	RedisModule_SaveUnsigned(rdb, v->type);
+	switch(v->type) {
+	case T_BOOL:
+	case T_INT64:
+		RedisModule_SaveSigned(rdb, v->longval);
+		return;
+	case T_DOUBLE:
+		RedisModule_SaveDouble(rdb, v->doubleval);
+		return;
+	case T_STRING:
+		RedisModule_SaveStringBuffer(rdb, v->stringval, strlen(v->stringval) + 1);
+		return;
+	case T_ARRAY:
+		_RdbSaveSIArray(rdb, *v);
+		return;
+	case T_NULL:
+		return; // No data beyond the type needs to be encoded for a NULL value.
+	default:
+		assert(0 && "Attempted to serialize value of invalid type.");
+	}
+}
+
+static void _RdbSaveSIArray(RedisModuleIO *rdb, const SIValue list) {
+	/* saves array as
+	   unsinged : array legnth
+	   array[0]
+	   .
+	   .
+	   .
+	   array[array length -1]
+	 */
+	uint arrayLen = SIArray_Length(list);
+	RedisModule_SaveUnsigned(rdb, arrayLen);
+	for(uint i = 0; i < arrayLen; i ++) {
+		SIValue value = SIArray_Get(list, i);
+		_RdbSaveSIValue(rdb, &value);
+	}
+}
+
+static void _RdbSaveEntity(RedisModuleIO *rdb, const Entity *e) {
+	/* Format:
+	 * #attributes N
+	 * (name, value type, value) X N  */
+
+	RedisModule_SaveUnsigned(rdb, e->prop_count);
+
+	for(int i = 0; i < e->prop_count; i++) {
+		EntityProperty attr = e->properties[i];
+		RedisModule_SaveUnsigned(rdb, attr.id);
+		_RdbSaveSIValue(rdb, &attr.value);
+	}
+}
+
+static void _RdbSaveEdge(RedisModuleIO *rdb, const Graph *g, const Edge *e, int r) {
+
+	/* Format:
+	* edge
+	* {
+	*  edge ID
+	*  source node ID
+	*  destination node ID
+	*  relation type
+	* }
+	* edge properties */
+
+	RedisModule_SaveUnsigned(rdb, ENTITY_GET_ID(e));
+
+	// Source node ID.
+	RedisModule_SaveUnsigned(rdb, Edge_GetSrcNodeID(e));
+
+	// Destination node ID.
+	RedisModule_SaveUnsigned(rdb, Edge_GetDestNodeID(e));
+
+	// Relation type.
+	RedisModule_SaveUnsigned(rdb, r);
+
+	// Edge properties.
+	_RdbSaveEntity(rdb, e->entity);
+}
+
+// Update the next encoding phase if needed.
+static void _UpdatedEncodePhase(GraphContext *gc) {
+	// Check if NODES encodeding phase is done
```

Comment:
I believe we can simplify this logic by first checking what phase we are in currently, as we never want to move backwards to an earlier phase.
```
	EncodePhase current_phase = GraphEncodeContext_GetEncodePhase(gc->encoding_context);
	switch(current_phase) {
	case NODES:
		if(GraphEncodeContext_GetProccessedNodes(gc->encoding_context) < Graph_NodeCount(gc->g)) {
			// We're not done encoding nodes, do not change phases.
			return;
		}
		// All non-deleted nodes have been encoded.
		if(array_len(gc->g->nodes->deletedIdx) > 0) {
			// We have deleted nodes that must be encoded, update phase.
			GraphEncodeContext_SetEncodePhase(gc->encoding_context, DELETED_NODES);
		} else {
			// No nodes left and no deleted nodes, set phase to edges.
			GraphEncodeContext_SetEncodePhase(gc->encoding_context, EDGES);
		}
		break;
	case DELETED_NODES:
		[...]
```
Let me know if you disagree, or if I'm missing something here!

Reply:
Each phase shares the conditions of the next phase and adds one of its own, so it ends up to be:
```
	EncodePhase current_phase = GraphEncodeContext_GetEncodePhase(gc->encoding_context);
	switch(current_phase) {
	case NODES:
		// Check if NODES encodeding phase is done
		if(GraphEncodeContext_GetProccessedNodes(gc->encoding_context) == Graph_NodeCount(gc->g)) {
			// Check if there is a need to encoded deleted nodes, or skip to edges.
			if(array_len(gc->g->nodes->deletedIdx) >
			   GraphEncodeContext_GetProccessedDeletedNodes(gc->encoding_context)) {
				GraphEncodeContext_SetEncodePhase(gc->encoding_context, DELETED_NODES);
			} else {
				// No nodes or deleted nodes left. Set state to edges.
				GraphEncodeContext_SetEncodePhase(gc->encoding_context, EDGES);
				// Check if need to encode edges, deleted edges or skip tp schema.
				if(GraphEncodeContext_GetProccessedEdges(gc->encoding_context) == Graph_EdgeCount(gc->g)) {
					// Check if there is a need to encoded deleted edges, or skip to schema.
					if(array_len(gc->g->edges->deletedIdx) >
					   GraphEncodeContext_GetProccessedDeletedEdges(gc->encoding_context)) {
						GraphEncodeContext_SetEncodePhase(gc->encoding_context, DELETED_EDGES);
					} else {
						GraphEncodeContext_SetEncodePhase(gc->encoding_context, GRAPH_SCHEMA);
					}
				}
			}
		}
		break;
	case DELETED_NODES:
		// Check if there is a need to encoded deleted nodes, or skip to edges.
		if(array_len(gc->g->nodes->deletedIdx) >
		   GraphEncodeContext_GetProccessedDeletedNodes(gc->encoding_context)) {
			return;
		} else {
			// No nodes or deleted nodes left. Set state to edges.
			GraphEncodeContext_SetEncodePhase(gc->encoding_context, EDGES);
			// Check if need to encode edges, deleted edges or skip tp schema.
			if(GraphEncodeContext_GetProccessedEdges(gc->encoding_context) == Graph_EdgeCount(gc->g)) {
				// Check if there is a need to encoded deleted edges, or skip to schema.
				if(array_len(gc->g->edges->deletedIdx) >
				   GraphEncodeContext_GetProccessedDeletedEdges(gc->encoding_context)) {
					GraphEncodeContext_SetEncodePhase(gc->encoding_context, DELETED_EDGES);
				} else {
					GraphEncodeContext_SetEncodePhase(gc->encoding_context, GRAPH_SCHEMA);
				}
			}
		}
		break;
	case EDGES:
		// Check if need to encode edges, deleted edges or skip tp schema.
		if(GraphEncodeContext_GetProccessedEdges(gc->encoding_context) == Graph_EdgeCount(gc->g)) {
			// Check if there is a need to encoded deleted edges, or skip to schema.
			if(array_len(gc->g->edges->deletedIdx) >
			   GraphEncodeContext_GetProccessedDeletedEdges(gc->encoding_context)) {
				GraphEncodeContext_SetEncodePhase(gc->encoding_context, DELETED_EDGES);
			} else {
				GraphEncodeContext_SetEncodePhase(gc->encoding_context, GRAPH_SCHEMA);
			}
		}
		break;
	case DELETED_EDGES:
		// Check if there is a need to encoded deleted edges, or skip to schema.
		if(array_len(gc->g->edges->deletedIdx) >
		   GraphEncodeContext_GetProccessedDeletedEdges(gc->encoding_context)) {
			return;
		} else {
			GraphEncodeContext_SetEncodePhase(gc->encoding_context, GRAPH_SCHEMA);
		}

	default:
		assert(false && "Unkown encode phase");
	}
```